### ISSUE TITLE
add "Nightwind Horn"

### ISSUE THREAD

**Issue by Zackx087** [Specy/genshin-music#60]
Nightwind Horn is a Gadget obtained from the Iridescent Arataki Rockin' For Life Tour de Force of Awesomeness event. according to wiki fandom
The notes only have 14
![Item_Nightwind_Horn](https://github.com/Specy/genshin-music/assets/61741707/b5fc1b5d-b355-4a5c-af37-cd9dee95d72f)

**Comment by ptr2019** [comment]
Just spotted this issue. In case the owner is too busy to answer in time, the following information is shared for your interest: As https://github.com/Specy/genshin-music/pull/59#issuecomment-2123545549 stated, the implement for the new instrument is on the list.

**Comment by Specy** [comment]
I didn't forget, it will just take a huge amount of time to add sustained notes to the app, that's why I'm not releasing it yet. I could try to add it without sustain in the meanwhile and then add it in the future

**Comment by ptr2019** [comment]
Non sustained is fine for now:laughing:

**Comment by ptr2019** [comment]
> I didn't forget, it will just take a huge amount of time to add sustained notes to the app, that's why I'm not releasing it yet. I could try to add it without sustain in the meanwhile and then add it in the future.
> 
> I'll make a poll in the community to see what they prefer, if either adding non sustained instrument for now, and then update in the future, or wait few months for the proper one to be released

May I ask how is the poll going?

**Comment by Specy** [comment]
> May I ask how is the poll going?

- 4 people want it to be added now as non sustained 

- 17 people want to wait for it to be added as sustained

**Comment by ptr2019** [comment]
Weeks have past~

May I ask something about the time schedule (e.g., is the framework ready now, with modules being implemented for those days).

**Comment by Specy** [comment]
I haven't had time to work on the app yet, I'm busy with work and university exams, (and I'm anxious about implementing this feature as it's huge work and I don't even know If it's gonna work)

**Comment by ptr2019** [comment]
Got it. Work and study should be in the first place. As for the feature, just take your time (a huge and different work indeed).

**Comment by ptr2019** [comment]
（A month later）

**Comment by Specy** [comment]
I know, i'm thinking of it everyday, i have the "last exam" of the month in a few days, after that i have a bit more free time to work on the app (hopefully)

**Comment by ptr2019** [comment]
No worry :smile: Just to let you know that there are still ones concerning about the issue～This work would be meaningful.

**Comment by ptr2019** [comment]
May I ask how is it going🙃(once again, awa)

**Comment by Specy** [comment]
Hello, I was in a bad burnout this month so i wasn't able to work on it. 

I'm currently waiting for https://github.com/pixijs/pixi-react/issues/493
To be shipped, I need to update the codebase to next.js 14 (hopefully by the time pixi react updates, next.js 15 comes out) and react 18 (or 19), then I can work on adding tone.js for the audio part that's needed for sustained instruments. I'll also talk to the people at tone.js to know the best approach to implement this

**Comment by ptr2019** [comment]
> Hello, I was in a bad burnout this month so i wasn't able to work on it.
> 
> I'm currently waiting for [pixijs/pixi-react#493](https://github.com/pixijs/pixi-react/issues/493) To be shipped, I need to update the codebase to next.js 14 (hopefully by the time pixi react updates, next.js 15 comes out) and react 18 (or 19), then I can work on adding tone.js for the audio part that's needed for sustained instruments. I'll also talk to the people at tone.js to know the best approach to implement this

Got it.

**Comment by cangerjun** [comment]
(three month later)

**Comment by ptr2019** [comment]
> (three month later)

I originally thought finally there was an update now, lol.

**Comment by cangerjun** [comment]
> > （三个月后）
> 
> 我原本以为现在终于有更新了，哈哈。

XD

**Comment by ptr2019** [comment]
Early New Year~ 
Any future plan :triangular_flag_on_post: for this project?

**Comment by cangerjun** [comment]
It's been two months since the official release of Next.js 15, are you OK?

**Comment by Specy** [comment]
https://github.com/pixijs-userland/graphics-smooth/issues/44

https://github.com/pixijs/pixi-react

Last two things needed...

**Comment by cangerjun** [comment]
> [pixijs-userland/graphics-smooth#44](https://github.com/pixijs-userland/graphics-smooth/issues/44)
> 
> https://github.com/pixijs/pixi-react
> 
> Last two things needed...
> 
> > are you OK?
> 
> Not really

Then you're busy, and you have two new instruments, a drum and a guitar with its own chords

**Comment by cangerjun** [comment]
Peace be upon you, Amen!

**Comment by ptr2019** [comment]
> [pixijs-userland/graphics-smooth#44](https://github.com/pixijs-userland/graphics-smooth/issues/44)
> 
> https://github.com/pixijs/pixi-react
> 
> Last two things needed...
> 
> > are you OK?
> 
> Not really

Hi, how is it now?

**Comment by Specy** [comment]
They are both not done yet

**Comment by ptr2019** [comment]
> They are both not done yet

I see. Just checked that these two. No update, nor further response.

**Comment by Specy** [comment]
Seems like https://github.com/pixijs/pixi-react is getting along, they are getting ready to publish the new version, not much info from graphics-smooth

**Comment by ptr2019** [comment]
Hi, any update? Note that discussion has been made in depth with other developer as above.

**Comment by Specy** [comment]
> Note that discussion has been made in depth with another developer as above.

What do you mean?

**Comment by ptr2019** [comment]
> > Note that discussion has been made in depth with another developer as above.
> 
> What do you mean?

This issue has been quoted in another issue https://github.com/Specy/genshin-music/issues/92 and the realization of long-press functionality is discussed there. 

Any update on this feature, i.e. Nightwind horn?

**Comment by Specy** [comment]
That issue was to ask information about the format, which referenced back this issue regarding the implementation.

I will update this issue whenever i start working on the feature in the future, i did not forget about it. So far no updates but i know of some people who are attempting sustained instruments in their application so i'm also gathering information from them about how it went so when i get the time to work on it i can make it properly.

**Comment by ptr2019** [comment]
Got that. Again, thanks for your attempt.

**Comment by Zackx087** [comment]
11 month later still not done yet😔

**Comment by ptr2019** [comment]
> 10 month later still not done yet😔

I figure that some dependencies for the implementation of sustained notes is still to be done. BTW I am wondering how would people want it to be for now, e.g., added the instrument as the non sustained version or wait for the sustained version.

**Comment by cangerjun** [comment]
> > 10个月过去了，仍然没有完成😔
> 
> 我觉得延音符实现的一些依赖关系还需要解决。顺便问一下，我想知道大家现在会怎么做，比如把乐器加成非延音版本，还是等延音版本。

已经制作了非延音版本，除了晚风圆号我依然没有任何头绪

**Comment by cangerjun** [comment]
@Specy Please see the last time I submitted the attached message, I will no longer start a new issues about this.

**Comment by Specy** [comment]
@cangerjun what do you mean?

**Comment by cangerjun** [comment]
> [@cangerjun](https://github.com/cangerjun)什么意思？

请在src\i18n\locales\index.ts中加入composer_note_name_type的翻译接口谢谢
这些内容我已加入zh.json
“composer_note_name_type”：{
“Note_name”： “音名”，
“Keyboard_layout”： “键盘布局”，
“Your_Keyboard_layout”： “你键盘的布局”，
"Do_Re_Mi": "哆 来 咪",
“ABC”：“ABC”，
"No_Text" : "无文本"
},

**Comment by cangerjun** [comment]
刚才的提交是：
1.修复sky的do re mi部分遗留错误，和1 2 3的配套错误
2.NumberLayout显示名称太长，改为‘1 2 3’ ，就像是’Do Re Mi‘这样的简单显示

**Comment by Specy** [comment]
I'm experimenting with sustained instruments in the Dev branch, i've rewritten the whole app in sveltekit instead of next.js since it was giving me too much pain to maintain, and i've implemented sustained notes and the ability to implement new games more easily (this is still to be done). I've done a ton of changes which need a lot of testing to make sure everything works (i have literally rewritten the whole app from scratch in a different language)

I've published the first sustained instrument in https://beta.sky-music.specy.app/ it is the "test sustain" which is not an actual sustained instrument, it is a bit bad as i dont have the audio files for proper sustained instruments so those need to be added afterwards (which i'd need some help with) 

Please test things out and tell me what you think

**Comment by Specy** [comment]
Ok i've added a proper sustained instrument example, it's called "Sustained recorder" in the Sky app (look at above message), which can be experimented with, afterwards i need the audio files for the nightwind horn and any other instrument which are sustained, and any other instrument that needs to be added to genshin music.

Sustained audio files are split in 3 parts, [attack, loop, end] attack and end are fixed for any length of sustain and are what is played when the note is just "pressed", the loop section is the part that repeats infinitely, for audio assets to be used in the app, there can be either two kinds of assets:

- the original audio asset where the audio is made up of exactly one attack, loop and end, the loop part has to go from start to end only ONCE, if it repeats then it won't be usable as is
- an audio sample of the instrument being used while sustained. If i have a long audio sample (like 10/20 seconds for each note) i can analyze the note itself to create the original attack, loop, end audio file and use that in the app (obviously there has to be no background noise or reverb, only the note)

So either the original audio assets with exactly one full loop, or a recorded audio sample with multiple loops in it so i can understand how to cut it

@ptr2019 @Zackx087 @cangerjun @QQATK

**Comment by Specy** [comment]
Nightwind horn has been added to beta https://beta.sky-music.specy.app/. I've also updated the other sustained instruments that were not implemented correctly. Please try them out and let me know what you think
